skiko/src/awtMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.kt

```diff
@@ -58,16 +62,13 @@ internal class MetalRedrawer(
         }
 
     init {
-        val adapter = chooseAdapter(properties.adapterPriority.ordinal)
-        val adapterName = getAdapterName(adapter)
-        val adapterMemorySize = getAdapterMemorySize(adapter)
-        onDeviceChosen(adapterName)
+        val adapter = chooseMetalAdapter(properties.adapterPriority)
+        onDeviceChosen(adapter.name)
         val initDevice = layer.backedLayer.useDrawingSurfacePlatformInfo {
-            MetalDevice(createMetalDevice(layer.windowHandle, layer.transparency, adapter, it))
+            MetalDevice(createMetalDevice(layer.windowHandle, layer.transparency, adapter.ptr, it))
```

Comment:
`createMetalDevice` releases the hard reference to `adapter.ptr` (calling __bridge_transfer), and after this function it is not valid accessing `adapter`.

Does calling `adapter.name` work after `createMetalDevice`?

To fix the issue, we need to replace `__bridge_transfer` by `__bridge` inside `createMetalDevice`, and create `MetalAdapter.dispose` that releases the pointer via `__bridge_transfer`.

__bridge_retained - acquires the ownership, we use it to pass to java
__bridge_transfer - releases the ownership, we use it to pass from java
__bridge - no changing the ownership, just convert it for passing from/to java. the object itself usually stored in an Objective C field.

(more in [this article](https://medium.com/@notestomyself/bridge-vs-bridge-transfer-vs-bridge-retained-f84e6b6406d1))

Reply:
> Does calling adapter.name work after createMetalDevice?

But it works fine and this code was written before, or do I miss something?

Reply:
We didn't use `adapter.ptr` after `createMetalDevice`. We used before it.

Reply:
But this is the same, since `adapter` is just a wrapper for `ptr`, `name`, and `memorySize`